sbndcode/OpDetReco/OpDeconvolution/SBNDOpDeconvolution_module.cc

```
@@ -0,0 +1,104 @@
+#include "messagefacility/MessageLogger/MessageLogger.h"
+
+#include <memory>
+
+#include "lardataobj/RawData/OpDetWaveform.h"
+
+#include "sbndcode/OpDetSim/sbndPDMapAlg.hh"
+#include "sbndcode/OpDetReco/OpDeconvolution/Alg/OpDeconvolutionAlg.hh"
+
+namespace opdet {
+  class SBNDOpDeconvolution;
+}
+
+
+class opdet::SBNDOpDeconvolution : public art::EDProducer {
+public:
+  explicit SBNDOpDeconvolution(fhicl::ParameterSet const& p);
+  // The compiler-generated destructor is fine for non-base
+  // classes without bare pointers or other resource use.
+
+  // Plugins should not be copied or assigned.
+  SBNDOpDeconvolution(SBNDOpDeconvolution const&) = delete;
+  SBNDOpDeconvolution(SBNDOpDeconvolution&&) = delete;
+  SBNDOpDeconvolution& operator=(SBNDOpDeconvolution const&) = delete;
+  SBNDOpDeconvolution& operator=(SBNDOpDeconvolution&&) = delete;
+
+  // Required functions.
+  void produce(art::Event& e) override;
+
+private:
+
+  // Declare member data here.
+  std::string fInputLabel;
+  std::vector<std::string> fPDTypes;
+  //OpDecoAlg
+  opdet::OpDeconvolutionAlg *fOpDecoAlg;
+  //PDS map
+  opdet::sbndPDMapAlg pdsmap;
+};
+
+
+opdet::SBNDOpDeconvolution::SBNDOpDeconvolution(fhicl::ParameterSet const& p)
+  : EDProducer{p}  // ,
+  // More initializers here.
+{
+  // Call appropriate produces<>() functions here.
+  // Call appropriate consumes<>() for any products to be retrieved by this module.
+  fInputLabel = p.get< std::string >("InputLabel");
+  fPDTypes = p.get< std::vector<std::string> >("PDTypes");
+  auto const opdeco_pset = p.get< fhicl::ParameterSet >("OpDecoAlg");
+  fOpDecoAlg = new opdet::OpDeconvolutionAlg(opdeco_pset);
+
+  produces< std::vector< raw::OpDetWaveform > >();
+}
+
+void opdet::SBNDOpDeconvolution::produce(art::Event& e)
+{
+  //Load the waveforms 
+  art::Handle< std::vector< raw::OpDetWaveform > > wfHandle;
+  e.getByLabel(fInputLabel, wfHandle);
+  if (!wfHandle.isValid()) {
+   std::cout<<"Non valid waveform handle\n";
```

cout statement ok, but think you should throw an error if it has bad input (maybe art will do this already since you declared produces? but would suggest guaranteeing it throws an error with a useful message)

---

Ah, sorry, I meant throw an exception too. Like: 

```
throw cet::exception("SBNDOpDeconvolution") << "Input waveforms with input label " << fInputLabel << " not found";
```